lte/gateway/python/magma/subscriberdb/crypto/milenage.py

```diff
@@ -54,6 +54,35 @@ def generate_eutran_vector(self, key, opc, sqn, plmn):
         kasme = Milenage.generate_kasme(ck, ik, plmn, sqn_bytes, ak)
         return rand, xres, autn, kasme
 
+    def generate_ngran_vector(self, key, opc, sqn, snni):
+        """
+        Generate the NGRAN key vector.
+        Args:
+            key (bytes): 128 bit subscriber key
+            opc (bytes): 128 bit operator variant algorithm configuration field
+            sqn (int): 48 bit sequence number
+            snni (bytes): 32 bit serving network name consisting of MCC and MNC
+        Returns:
+            rand (bytes): 128 bit random challenge
+            xres (bytes): 128 bit expected result
+            autn (bytes): 128 bit authentication token
+            kseaf (bytes): 256 bit base network authentication code
+        """
+        sqn_bytes = bytearray.fromhex('{:012x}'.format(sqn))
+        rand = Milenage.generate_rand()
+
+        mac_a, _ = Milenage.f1(key, sqn_bytes, rand, opc, self.amf)
+        xres, ak = Milenage.f2_f5(key, rand, opc)
+        ck = Milenage.f3(key, rand, opc)
+        ik = Milenage.f4(key, rand, opc)
+
+        autn = Milenage.generate_autn(sqn_bytes, ak, mac_a, self.amf)
+        xres_star = Milenage.generate_m5g_xres_star(ck + ik, snni, rand, xres)
+        kausf = Milenage.generate_m5g_kausf(ck + ik, snni, autn)
+        kseaf = Milenage.generate_m5g_kseaf(kausf, snni)
+
+        return rand, xres_star, autn, kseaf
```

Comment:
same here, let's convert these to a NamedTuple

Reply:
I've addressed the comments.